src/module_handle.cc

```diff
@@ -0,0 +1,345 @@
+	//	return ptr == data.second;
+	//});
+}
+
+
+IsolatedModule::IsolatedModule(std::shared_ptr<IsolateHolder> isolate, std::shared_ptr<RemoteHandle<v8::Module>> handle, std::vector<std::string> dependencySpecifiers)
+	: isolate(std::move(isolate)), module_handle(std::move(handle)), dependencySpecifiers(std::move(dependencySpecifiers))
+{
+}
+
+
+void IsolatedModule::lock()
+{
+	return mutex.lock();
+}
+
+void IsolatedModule::unlock()
+{
+	return mutex.unlock();
+}
+
+
+const std::vector<std::string> & IsolatedModule::GetDependencySpecifiers() const
+{
+	return dependencySpecifiers;
+}
+
+void IsolatedModule::SetDependency(std::string specifier, std::shared_ptr<IsolatedModule> isolated_module) {
+	// Probably not a good idea because if dynamic import() anytime get supported()
+	std::lock_guard<IsolatedModule> lock(*this);
+	if (std::find(GetDependencySpecifiers().begin(), GetDependencySpecifiers().end(), specifier) == GetDependencySpecifiers().end()) {
+	  const std::string errorMessage = "Module has no dependency named: \"" + specifier + "\".";
+	  throw js_generic_error(errorMessage.c_str());
+	}
+	resolutions[specifier] = isolated_module;
+}
+
+
+
+MaybeLocal<Module> IsolatedModule::ResolveCallback(Local<Context> context, Local<String> specifier, Local<Module> referrer) {
+	std::string dependency = *Utf8ValueWrapper(Isolate::GetCurrent(), specifier);
+	IsolatedModule* ptr = shared::find(referrer);
+	if (ptr != nullptr) {
+		shared_ptr<IsolatedModule> found = ptr->shared_from_this();
+		
+		// I don�t believe we must check the context, but I am unsure
+		if (found->context_handle->Deref() != context) {
+			throw js_generic_error((std::string("Failed to resolve dependency: \"") + dependency + std::string("\". Invalid context")).c_str());
+		}
+
+		std::lock_guard<IsolatedModule> lock(*found);
+		auto & resolutions = found->resolutions;
+		auto it = resolutions.find(dependency);
+		
+
+		if (it != resolutions.end()) {
+			return it->second->module_handle->Deref();
+		}
+	}
+	throw js_generic_error((std::string("Failed to resolve dependency: ") + dependency).c_str());
+}
+
+
+void IsolatedModule::Instantiate(std::shared_ptr<RemoteHandle<v8::Context>> c) {
+	std::lock_guard<IsolatedModule> lock(*this);
+	context_handle = c;
+	Isolate* isolate = Isolate::GetCurrent();
+	Local<Context> context = context_handle->Deref();
+	Local<Module> mod = module_handle->Deref();
+	bool result = Unmaybe(mod->InstantiateModule(context, IsolatedModule::ResolveCallback));
+	if (!result) {
+		throw js_generic_error("Failed to instantiate module");
+	}
+}
+
+std::unique_ptr<Transferable> IsolatedModule::Evaluate(std::size_t timeout) {
+	std::lock_guard<IsolatedModule> lock(*this);
+	Local<Context> context_local = Deref(*context_handle);
+	Context::Scope context_scope(context_local);
+	Local<Module> mod = module_handle->Deref();
+	return ExternalCopy::CopyIfPrimitive(RunWithTimeout(timeout, [&]() {
```

Comment:
This should just be:
```c++
  std::unique_ptr<Transferable> ret = ExternalCopy::CopyIfPrimitive(
    RunWithTimeout(timeout, [&]() { return mod->Evaluate(context_local); })
  );           
  global_namespace = std::make_shared<RemoteHandle<Value>>(mod->GetModuleNamespace());
  return ret;
```

The reason this was failing is because `RunWithTimeout` causes v8 to terminate execution, and then `RunWithTimeout` will throw an exception. The exception can't be thrown until `RunWithTimeout` finishes though, and `Evaluate` returns before finishing. By moving the call to `GetModuleNamespace()` out of the timeout you can be assured that `Evaluate` completed successfully.

You had a good unit test here that caught the error 👍🏻

Reply:
Fixed.